File: BFS/987.vertical-order-traversal-of-a-binary-tree.go
package leetcode

import "sort"

// @leet start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type node1 struct {
	*TreeNode
	col, row int
}

func verticalTraversal(root *TreeNode) [][]int {
	var (
		l, r int
		ret  [][]int
	)
	mp := map[int][]node1{}
	queue := []node1{{root, 0, 0}}

	for len(queue) > 0 {
		top := queue[0]
		queue = queue[1:]
		mp[top.col] = append(mp[top.col], top)
		l = min(l, top.col)
		r = max(r, top.col)
		if top.Left != nil {
			queue = append(queue, node1{top.Left, top.col - 1, top.row + 1})
		}
		if top.Right != nil {
			queue = append(queue, node1{top.Right, top.col + 1, top.row + 1})
		}
	}
	for i := l; i <= r; i++ {
		cur := mp[i]
		sort.Slice(cur, func(i, j int) bool {
			l, r := cur[i], cur[j]
			if l.row != r.row {
				return l.row < r.row
			}
			return l.TreeNode.Val < r.TreeNode.Val
		})
		temp := make([]int, len(cur))
		for i, v := range cur {
			temp[i] = v.TreeNode.Val
		}
		ret = append(ret, temp)
	}

	return ret
}

// @leet end

func verticalTraversal_2(root *TreeNode) [][]int {
	type item struct {
		c int
		n *TreeNode
	}
	type n struct {
		r, v int
	}
	var (
		minC, maxC, level int
		queue             = []item{{0, root}}
		m                 = map[int][]n{}
	)
	for len(queue) > 0 {

		for range len(queue) {
			cur := queue[0]
			queue = queue[1:]
			maxC, minC = max(cur.c, maxC), min(cur.c, minC)
			m[cur.c] = append(m[cur.c], n{level, cur.n.Val})
			if cur.n.Left != nil {
				queue = append(queue, item{cur.c - 1, cur.n.Left})
			}
			if cur.n.Right != nil {
				queue = append(queue, item{cur.c + 1, cur.n.Right})
			}
		}
		level++
	}
	ret := make([][]int, 0, maxC-minC+1)
	for i := minC; i <= maxC; i++ {
		// ret[i] = m[i] // have negative index
		curC := m[i]
		sort.Slice(curC, func(i, j int) bool {
			if curC[i].r == curC[j].r {
				return curC[i].v < curC[j].v
			}
			return curC[i].r < curC[j].r
		})
		temp := make([]int, len(curC))
		for i, v := range curC {
			temp[i] = v.v
		}
		ret = append(ret, temp)
	}
	return ret
}

File: BFS/987.vertical-order-traversal-of-a-binary-tree.py
from typing import DefaultDict, List, Optional
from type import TreeNode
from collections import defaultdict, deque


# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from dataclasses import dataclass


# NOTE: use dataclass as struct to simplify code
@dataclass(slots=True)
class item:
    col: int
    row: int
    node: TreeNode


class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        maxC = minC = 0
        colM: DefaultDict[int | str, list[item]] = defaultdict(list)
        q = deque([item(0, 0, root)])
        while len(q) > 0:
            cur = q.popleft()
            maxC = max(maxC, cur.col)
            minC = min(minC, cur.col)
            colM[cur.col].append(cur)
            if cur.node.left is not None:
                q.append(item(cur.col - 1, cur.row + 1, cur.node.left))
            if cur.node.right is not None:
                q.append(item(cur.col + 1, cur.row + 1, cur.node.right))

        res = []
        for i in range(minC, maxC + 1):
            cur = colM[i]
            cur.sort(key=lambda x: (x.row, x.node.val))
            res.append([v.node.val for v in cur])
        return res


# @leet end
